src/ompl/base/samplers/DeterministicStateSampler.h

```diff
@@ -0,0 +1,172 @@
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the Rice University nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Leonard Bruns */
+
+#ifndef OMPL_BASE_DETERMINISTIC_STATE_SAMPLER_
+#define OMPL_BASE_DETERMINISTIC_STATE_SAMPLER_
+
+#include "ompl/base/StateSpace.h"
+#include "ompl/base/StateSampler.h"
+#include "ompl/base/samplers/deterministic/DeterministicSequence.h"
+
+#include <memory>
+
+namespace ompl
+{
+    namespace base
+    {
+        /**
+       @anchor DeterministicStateSampler
+       @par Short description
+       \ref DeterministicStateSampler Implementation of a deterministic state sampler. The
+       implementation allows to load and draw samples from a precomputed sequence or from the Halton Sequence,
+       @par External documentation
+       Dispertio: Optimal Sampling For Safe Deterministic Motion Planning
+       L Palmieri, L Bruns, M Meurer, KO Arras - IEEE Robotics and Automation Letters, 2019
+       DOI: [10.1109/LRA.2019.2958525](https://ieeexplore.ieee.org/abstract/document/8928532)<br>
+       [[PDF]](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8928532)
+    */
+
+        /** \brief An abstract class for the concept of using deterministic sampling sequences
+        to decrease the dispersion of the samples. */
+        class DeterministicStateSampler : public StateSampler
+        {
+        public:
+            enum DeterministicSamplerType
+            {
+                HALTON
```

Comment:
It'd nice if there was support for at least one more deterministic sampler (e.g., Hammersley). This would help establish that the API is general enough to accommodate other types.

Why isn't there a PrecomputedStateSampler type?

Reply:
I'm the original author of this, but not at Bosch anymore, so I am not in the position to address the changes directly. Just wanted to comment on this. I believe the problem with the PrecomputedStateSampler type would be that there is no suitable default constructor that works out of the box (i.e., you need to specify the path for it to make sense).

So for the PrecomputedSequence you'd have to use the sequence_ptr constructor instead. Alternative would be to change the PrecomputedSequence a bit and allow access to sequence_ptr_.

Reply:
I almost forgot about this, but OMPL already has a class for sampling precomputed samples. See `ompl::base::StateStorage` and `tests/base/state_storage.cpp`. Would it be possible to eliminate the PrecomputedStateSampler class in favor of using [StateStorage::getStateSamplerAllocator](https://ompl.kavrakilab.org/classompl_1_1base_1_1StateStorage.html#a0acdbdeeb523a654afb94b9c2de9b6eb)?

Reply:
We would love to keep the PrecomputedSequence class. This would allow us to include easily new functionalities, for instance a method that handles the shuffling of the samples.